pyomo/contrib/mindtpy/cut_generation.py

```diff
@@ -0,0 +1,99 @@
+"""Cut generation."""
+from __future__ import division
+
+from math import copysign, fabs
+
+from pyomo.core import Constraint, Var, minimize, value
+from pyomo.core.expr import current as EXPR
+from pyomo.core.expr.current import ExpressionReplacementVisitor
+from pyomo.repn import generate_standard_repn
+from pyomo.core.kernel.component_set import ComponentSet
+
+
+def add_objective_linearization(solve_data, config):
+    """Adds initial linearized objective in case it is nonlinear.
+
+    This should be done for initializing the ECP method.
+
+    """
+    m = solve_data.working_model
+    MindtPy = m.MindtPy_utils
+    solve_data.mip_iter += 1
+    gen = (obj for obj in MindtPy.jacs
+           if obj is MindtPy.MindtPy_objective_expr)
+    MindtPy.MindtPy_linear_cuts.mip_iters.add(solve_data.mip_iter)
+    sign_adjust = 1 if MindtPy.obj.sense == minimize else -1
+    # generate new constraints
+    # TODO some kind of special handling if the dual is phenomenally small?
+    for obj in gen:
+        c = MindtPy.MindtPy_linear_cuts.ecp_cuts.add(
+            expr=sign_adjust * sum(
+                value(MindtPy.jacs[obj][id(var)]) * (var - value(var))
+                for var in list(EXPR.identify_variables(obj.body))) +
+            value(obj.body) <= 0)
+        MindtPy.ECP_constr_map[obj, solve_data.mip_iter] = c
+
+
+def add_oa_cut(var_values, duals, solve_data, config):
+    m = solve_data.mip
+    MindtPy = m.MindtPy_utils
+    MindtPy.MindtPy_linear_cuts.nlp_iters.add(solve_data.nlp_iter)
+    sign_adjust = -1 if solve_data.objective_sense == minimize else 1
+
+    # Copy values over
+    for var, val in zip(MindtPy.variable_list, var_values):
+        if val is not None and not var.fixed:
+            var.value = val
+
+    # generate new constraints
+    # TODO some kind of special handling if the dual is phenomenally small?
+    jacs = solve_data.jacobians
+    for constr, dual_value in zip(MindtPy.constraint_list, duals):
+        if constr.body.polynomial_degree() in (1, 0):
+            continue
+        rhs = ((0 if constr.upper is None else constr.upper) +
+               (0 if constr.lower is None else constr.lower))
```

Comment:
You may want to consider the case of `constr.upper = constr.lower = some number`.

Reply:
After some thought, I have decided to subtract the lower bound if both exist. In case of an equality constraint, we will result in an RHS equal to the original RHS and in the next lines relaxed to be generated in whichever direction the dual says. If it has both bounds and are different, it uses as the RHS the upper bound, which will be looser and therefore less risky to cut stuff that we wouldn't like to cut (UB>=LB).
Solved in the last commit.